File: services/rust/axum/src/routes/middleware/chaos.rs
use crate::routes::errors::AppError;
use app::state::AppState;
use axum::{
    extract::{Request, State},
    middleware::Next,
    response::{IntoResponse, Response},
};
use rand::{RngExt, rng};
use std::{
    sync::{Arc, Mutex},
    time::Duration,
};
use tokio::time::sleep;
use tracing::instrument;

#[instrument(skip_all)]
pub async fn chaos_middleware(
    State(state): State<Arc<Mutex<AppState>>>,
    req: Request,
    next: Next,
) -> Response {
    let (latency_enabled, latency_rate, error_enabled, error_rate) = match state.lock() {
        Ok(guard) => (
            guard.config.latency.enabled,
            guard.config.latency.rate,
            guard.config.error.enabled,
            guard.config.error.rate,
        ),
        Err(_) => {
            tracing::error!(
                method = %req.method(),
                path = %req.uri(),
                "Poisoned lock in chaos_middleware"
            );
            return AppError::Internal.into_response();
        }
    };

    if latency_enabled && rng().random_range(0..=100) < latency_rate {
        let duration = Duration::from_micros(rng().random_range(500..=1500));
        sleep(duration).await;
    }

    if error_enabled && rng().random_range(0..=100) < error_rate {
        return AppError::Internal.into_response();
    }

    next.run(req).await
}

File: services/go/fiber/routes/middleware/chaos.go
package middleware

import (
	"app"
	"fiber/routes"
	"math/rand/v2"
	"time"

	"github.com/gofiber/fiber/v3"
)

func ChaosMiddleware(state *app.AppState) fiber.Handler {
	return func(c fiber.Ctx) error {
		latencyEnabled := state.Config.Latency.Enabled
		latencyRate := state.Config.Latency.Rate
		errorEnabled := state.Config.Error.Enabled
		errorRate := state.Config.Error.Rate

		if latencyEnabled && rand.IntN(101) < int(latencyRate) {
			delay := time.Duration(500+rand.IntN(1001)) * time.Microsecond
			time.Sleep(delay)
		}
		if errorEnabled && rand.IntN(101) < int(errorRate) {
			return routes.AppErrors.Internal.Error(c)
		}

		return c.Next()
	}
}
